File: scripts/publish-telegraph.py
#!/usr/bin/env python3
"""
Publish an English MDX blog post to Telegraph (telegra.ph).
The full article is published (not an announcement), with a link
to the original Russian version appended at the end.

Usage:
    python publish-telegraph.py <en_mdx_file> <ru_slug>

Environment variables:
    TELEGRAPH_ACCESS_TOKEN  – Telegraph account access token (required)
    SITE_BASE_URL           – Base URL of the site (default: https://iconicompany.com)
"""

import json
import os
import re
import sys
from html.parser import HTMLParser

import markdown
import requests

# Telegraph only supports a limited set of tags.
ALLOWED_TAGS = {
    "a", "aside", "b", "blockquote", "br", "caption", "code", "del",
    "details", "div", "em", "figcaption", "figure", "footer", "h3", "h4",
    "hr", "i", "iframe", "img", "kbd", "li", "mark", "ol", "p", "pre",
    "s", "section", "strike", "strong", "sub", "sup", "table", "td", "th",
    "thead", "tbody", "tfoot", "time", "tr", "ul", "var", "video",
}

# Map unsupported heading tags to the nearest allowed equivalent.
TAG_MAP = {
    "h1": "h3",
    "h2": "h3",
    "h5": "h4",
    "h6": "h4",
}


class HtmlToNodes(HTMLParser):
    """Parse an HTML string into a Telegraph Node array."""

    def __init__(self):
        super().__init__()
        self._root = []
        self._stack = [self._root]

    def handle_starttag(self, tag, attrs):
        tag = TAG_MAP.get(tag, tag)
        if tag not in ALLOWED_TAGS:
            return
        node = {"tag": tag}
        attrs_dict = dict(attrs)
        if tag == "a" and "href" in attrs_dict:
            node["attrs"] = {"href": attrs_dict["href"]}
        elif tag == "img" and "src" in attrs_dict:
            node["attrs"] = {"src": attrs_dict["src"]}
            if "alt" in attrs_dict:
                node["attrs"]["alt"] = attrs_dict["alt"]
        node["children"] = []
        self._stack[-1].append(node)
        self._stack.append(node["children"])

    def handle_endtag(self, tag):
        tag = TAG_MAP.get(tag, tag)
        if tag in ALLOWED_TAGS and len(self._stack) > 1:
            self._stack.pop()

    def handle_data(self, data):
        if data:
            self._stack[-1].append(data)

    @property
    def nodes(self):
        return self._root


def html_to_nodes(html_text):
    parser = HtmlToNodes()
    parser.feed(html_text)
    return parser.nodes


def parse_mdx(filepath):
    """Return (frontmatter_str, body_str) from an MDX/MD file."""
    with open(filepath, encoding="utf-8") as fh:
        content = fh.read()
    match = re.match(r"^---\s*\n(.*?)\n---\s*\n(.*)$", content, re.DOTALL)
    if match:
        return match.group(1), match.group(2).strip()
    return "", content.strip()


def extract_title(frontmatter):
    for pattern in (
        r'^title:\s*"(.*?)"',
        r"^title:\s*'(.*?)'",
        r"^title:\s*(.+)",
    ):
        m = re.search(pattern, frontmatter, re.MULTILINE)
        if m:
            return m.group(1).strip().strip("'\"")
    return "Article"


def publish(access_token, title, nodes, author_name="Iconicompany", author_url="https://iconicompany.com"):
    response = requests.post(
        "https://api.telegra.ph/createPage",
        data={
            "access_token": access_token,
            "title": title,
            "author_name": author_name,
            "author_url": author_url,
            "content": json.dumps(nodes, ensure_ascii=False),
            "return_content": "false",
        },
        timeout=30,
    )
    response.raise_for_status()
    return response.json()


def main():
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} <en_mdx_file> <ru_slug>", file=sys.stderr)
        sys.exit(1)

    en_file = sys.argv[1]
    ru_slug = sys.argv[2]

    access_token = os.environ.get("TELEGRAPH_ACCESS_TOKEN", "")
    if not access_token:
        print("TELEGRAPH_ACCESS_TOKEN is not set, skipping.", file=sys.stderr)
        sys.exit(0)

    site_base = os.environ.get("SITE_BASE_URL", "https://iconicompany.com").rstrip("/")
    original_url = f"{site_base}/ru/blog/{ru_slug}"
    en_url = f"{site_base}/en/blog/{ru_slug}"

    frontmatter, body = parse_mdx(en_file)
    title = extract_title(frontmatter)

    html_body = markdown.markdown(
        body,
        extensions=["tables", "fenced_code"],
    )

    # Append links to the English and original Russian articles.
    original_link_html = (
        f'<p><a href="{en_url}">Read in English</a> | '
        f'<a href="{original_url}">Read original in Russian</a></p>'
    )
    full_html = html_body + "\n" + original_link_html

    nodes = html_to_nodes(full_html)

    result = publish(access_token, title, nodes)

    if result.get("ok"):
        url = result["result"]["url"]
        print(url)
    else:
        print(f"Telegraph API error: {result}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/publish-telegraph.ts
import { readFileSync, existsSync } from "fs";
import { marked } from "marked";

const SITE_BASE_URL = process.env.SITE_BASE_URL || "https://iconicompany.com";
const TELEGRAPH_ACCESS_TOKEN = process.env.TELEGRAPH_ACCESS_TOKEN || "";

type TelegraphNode = string | {
  tag: string;
  attrs?: Record<string, string>;
  children?: TelegraphNode[];
};

export function htmlToNodes(html: string): TelegraphNode[] {
  const nodes: TelegraphNode[] = [];
  const stack: TelegraphNode[][] = [nodes];
  
  const TAG_MAP: Record<string, string> = {
    h1: "h3",
    h2: "h3",
    h5: "h4",
    h6: "h4"
  };
  
  const ALLOWED_TAGS = new Set([
    "a", "aside", "b", "blockquote", "br", "caption", "code", "del",
    "details", "div", "em", "figcaption", "figure", "footer", "h3", "h4",
    "hr", "i", "iframe", "img", "kbd", "li", "mark", "ol", "p", "pre",
    "s", "section", "strike", "strong", "sub", "sup", "table", "td", "th",
    "thead", "tbody", "tfoot", "time", "tr", "ul", "var", "video"
  ]);

  const tagRegex = /(<([^>]+)>)/ig;
  let lastIndex = 0;
  let match;

  while ((match = tagRegex.exec(html)) !== null) {
    const textBetween = html.substring(lastIndex, match.index);
    if (textBetween) {
      const currentStack = stack[stack.length - 1];
      if (currentStack) {
        currentStack.push(textBetween);
      }
    }
    
    const tagToken = match[2];
    if (tagToken === undefined) continue;
    const isClosing = tagToken.startsWith("/");
    
    if (isClosing) {
      const tag = tagToken.substring(1).toLowerCase();
      const mappedTag = TAG_MAP[tag] || tag;
      if (ALLOWED_TAGS.has(mappedTag) && stack.length > 1) {
        stack.pop();
      }
    } else {
      const parts = tagToken.split(/\s+/);
      const parts0 = parts[0];
      if (parts0 !== undefined) {
        const originalTag = parts0.replace(/\/$/, "").toLowerCase();
        const isSelfClosing = tagToken.endsWith("/");
        
        const tag = TAG_MAP[originalTag] || originalTag;
        
        if (ALLOWED_TAGS.has(tag)) {
          const node: any = { tag, children: [] };
          
          // Parse attributes cleanly
          const attrRegex = /([a-zA-Z\-]+)="([^"]+)"/g;
          let attrMatch;
          const attrs: Record<string, string> = {};
          while ((attrMatch = attrRegex.exec(tagToken)) !== null) {
            const attrKey = attrMatch[1];
            const attrVal = attrMatch[2];
            if (attrKey !== undefined && attrVal !== undefined) {
              attrs[attrKey.toLowerCase()] = attrVal;
            }
          }
          
          if (tag === "a" && attrs.href) {
            node.attrs = { href: attrs.href };
          } else if (tag === "img" && attrs.src) {
            node.attrs = { src: attrs.src };
            if (attrs.alt) node.attrs.alt = attrs.alt;
          }
          
          const currentStack = stack[stack.length - 1];
          if (currentStack) {
            currentStack.push(node);
            if (!isSelfClosing && tag !== "br" && tag !== "hr" && tag !== "img") {
              stack.push(node.children);
            }
          }
        }
      }
    }
    
    lastIndex = tagRegex.lastIndex;
  }
  
  const remainingText = html.substring(lastIndex);
  if (remainingText) {
    const currentStack = stack[stack.length - 1];
    if (currentStack) {
      currentStack.push(remainingText);
    }
  }
  
  return nodes;
}

function parseMdx(filePath: string) {
  const content = readFileSync(filePath, "utf-8");
  const match = content.match(/^---\s*\n([\s\S]*?)\n---\s*\n([\s\S]*)$/);
  if (match) {
    const fm = match[1];
    const body = match[2];
    if (fm !== undefined && body !== undefined) {
      return { frontmatter: fm, body: body.trim() };
    }
  }
  return { frontmatter: "", body: content.trim() };
}

function extractTitle(frontmatter: string): string {
  const match = frontmatter.match(/^title:\s*["']?(.*?)["']?$/m);
  if (match) {
    const title = match[1];
    if (title !== undefined) return title.trim();
  }
  return "Article";
}

export async function publishTelegraph(options: {
  accessToken: string;
  title: string;
  nodes: TelegraphNode[];
  authorName?: string;
  authorUrl?: string;
}): Promise<string> {
  const { accessToken, title, nodes, authorName = "Iconicompany", authorUrl = "https://iconicompany.com" } = options;
  if (!accessToken || !title || !nodes) {
    throw new Error("Missing required options for Telegraph publishing.");
  }

  const formData = new FormData();
  formData.append("access_token", accessToken);
  formData.append("title", title);
  formData.append("author_name", authorName);
  formData.append("author_url", authorUrl);
  formData.append("content", JSON.stringify(nodes));
  formData.append("return_content", "false");

  const response = await fetch("https://api.telegra.ph/createPage", {
    method: "POST",
    body: formData
  });

  if (!response.ok) {
    const errText = await response.text();
    throw new Error(`Telegraph API failed: ${response.status} - ${errText}`);
  }

  const json: any = await response.json();
  if (json.ok) {
    return json.result.url;
  } else {
    throw new Error(`Telegraph API error: ${JSON.stringify(json)}`);
  }
}

function usage() {
  console.log("Usage: bun scripts/publish-telegraph.ts <en-mdx-file> <ru-slug>");
  process.exit(1);
}

async function main() {
  const args = Bun.argv.slice(2);
  const enFile = args[0];
  const ruSlug = args[1];

  if (!enFile || !ruSlug) {
    usage();
    return;
  }

  if (!TELEGRAPH_ACCESS_TOKEN) {
    console.warn("TELEGRAPH_ACCESS_TOKEN is not set, skipping.");
    process.exit(0);
  }

  if (!existsSync(enFile)) {
    console.error(`Error: File not found at '${enFile}'`);
    process.exit(1);
  }

  const siteBase = SITE_BASE_URL.replace(/\/$/, "");
  const originalUrl = `${siteBase}/ru/blog/${ruSlug}`;
  const enUrl = `${siteBase}/en/blog/${ruSlug}`;

  const { frontmatter, body } = parseMdx(enFile);
  const title = extractTitle(frontmatter);

  // Parse markdown
  const htmlBody = await marked.parse(body);

  const originalLinkHtml = `<p><a href="${enUrl}">Read in English</a> | <a href="${originalUrl}">Read original in Russian</a></p>`;
  const fullHtml = `${htmlBody}\n${originalLinkHtml}`;

  const nodes = htmlToNodes(fullHtml);

  try {
    const url = await publishTelegraph({
      accessToken: TELEGRAPH_ACCESS_TOKEN,
      title,
      nodes
    });
    console.log(url);
  } catch (err: any) {
    console.error("Fatal Error:", err.message);
    process.exit(1);
  }
}

if (import.meta.main) {
  main();
}
